File: projects/mcp-servers/csharp/Tools/CalculateTool.cs
using ModelContextProtocol;
using ModelContextProtocol.Server;
using System.ComponentModel;

namespace CSharpMcpServer.Tools;

[McpTool("calculate")]
[Description("Performs a basic arithmetic operation on two numbers.")]
public sealed class CalculateTool
{
    [McpToolParameter]
    [Description("The first operand.")]
    public double A { get; set; }

    [McpToolParameter]
    [Description("The second operand.")]
    public double B { get; set; }

    [McpToolParameter]
    [Description("The arithmetic operation to perform. Valid values: add, subtract, multiply, divide, power, modulo.")]
    public string Operation { get; set; } = "add";

    [McpToolMethod]
    [Description("Executes the calculation and returns the result.")]
    public string Execute()
    {
        var result = Operation.ToLowerInvariant() switch
        {
            "add" => A + B,
            "sum" => A + B,
            "subtract" => A - B,
            "minus" => A - B,
            "multiply" => A * B,
            "times" => A * B,
            "divide" or "divided by" when B == 0
                => throw new DivideByZeroException("Division by zero is not allowed."),
            "divide" => A / B,
            "divided by" => A / B,
            "power" => Math.Pow(A, B),
            "pow" => Math.Pow(A, B),
            "modulo" => A % B,
            "mod" => A % B,
            _ => throw new ArgumentException($"Unknown operation '{Operation}'. Valid operations: add, subtract, multiply, divide, power, modulo.")
        };

        return $"{A} {Operation} {B} = {result}";
    }
}

File: projects/mcp-servers/ruby/lib/my_mcp_server/tools/calculate_tool.rb
# frozen_string_literal: true

module MyMcpServer
  module Tools
    # A calculator tool supporting basic arithmetic.
    class CalculateTool
      OPERATIONS = %w[add subtract multiply divide].freeze

      def self.name
        'calculate'
      end

      def self.description
        'Perform basic arithmetic (add, subtract, multiply, divide).'
      end

      def self.schema
        {
          type: 'object',
          properties: {
            operation: {
              type: 'string',
              enum: OPERATIONS,
              description: 'The arithmetic operation to perform'
            },
            a: {
              type: 'number',
              description: 'First operand'
            },
            b: {
              type: 'number',
              description: 'Second operand'
            }
          },
          required: %w[operation a b]
        }
      end

      def self.execute(args)
        op = args['operation']
        a  = args['a'].to_f
        b  = args['b'].to_f

        result = case op
                 when 'add'      then a + b
                 when 'subtract' then a - b
                 when 'multiply' then a * b
                 when 'divide'
                   raise ArgumentError, 'Division by zero' if b.zero?

                   a / b
                 else
                   raise ArgumentError, "Unknown operation: #{op}"
                 end

        # Return integer result when both operands are whole numbers
        result = result.to_i if result == result.to_i

        { result: result, operation: op, a: a, b: b }
      end
    end
  end
end
